src/createBlock.ts

```diff
@@ -0,0 +1,147 @@
+/**
+ * Creates the next block in the Raha Blockchain from unapplied operations in Firestore.
+ *
+ * Example usage:
+ * node dist/createBlock.js > newBlock.json
+ */
+
+import { getBlockchain } from './RahaBlockchain';
+import { operationsCollectionFilters as filters, get, operationsCollection } from './RahaFirestore';
+import { BLOCKCHAIN_VERSION_NO, Block, StellarMetadata, IpfsBlock, Operation, RequestInvite, Trust } from './schema';
+
+/**
+ * A map of functions that build the data component of Blockchain Operations from
+ * the data component of Firestore Operations.
+ */
+const bcOpDataBuilders = {
+    'REQUEST_INVITE': (fsOpData): RequestInvite => ({
+        full_name: fsOpData.full_name,
+        to_mid: fsOpData.to_mid,
+        video_url: fsOpData.video_url,
+    }),
+    'TRUST': (operationData): Trust => ({
+        to_mid: operationData.to_mid,
+    }),
+}
+
+/**
+ * Filter new operations are invalid within the context of the current blockchain.
+ * This means:
+ *   * users cannot perform operations on nonexistent users.
+ *   * // TODO: there must exist a trust path from a user who joined in a previous block to any user who performs a new TRUST operation
+ */
+function filterBcOpsAgainstBlockchain(blockchain, bcOps) {
+    // Create a set of creator_mid on all REQUEST_INVITE operations in blockchain and bcOps.
+    // Check if to_mid, from_mid is included in above set.
+    // TODO;
+    return bcOps;
+}
+
+/**
+ * TODO.
+ */
+function validateBcOp(bcOp) {
+    return true;
+}
+
+/**
+ * Build a Blockchain Operation from a Firestore Operation. Returns undefined if the operation is invalid.
```

Comment:
Is it best for us to return `undefined` in a failure like this, or would it make more sense to throw a subclass of `Error`? It seems like providing an invalid firestore op is a misuse of the API, makes upstream code more explicit about the fact that this is an error case.

If you do go this path, I'd make one subclass of `Error` that is a `RahaBlockchainError`, and then if you need something more specific than that then subclass that further.

Note: Subclassing `Error` has a little weirdness, so [follow the directions here](https://github.com/Microsoft/TypeScript/wiki/FAQ#why-doesnt-extending-built-ins-like-error-array-and-map-work) to get it right.

Reply:
Hmm, long term I think that our blockchain client doesn't really care about whether it sees an operation with a op_code it doesn't recognize or not. All operations it receives are untrusted. It'll only include the ones that are valid according to it's rules, but I wouldn't say it's an "error" per se to have an invalid op.